lib/chef/provider/profile.rb

```diff
@@ -0,0 +1,253 @@
+#
+# Author:: Nate Walck (<nate.walck@gmail.com>)
+# Copyright:: Copyright (c) 2015 Facebook, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef/log'
+require 'chef/provider'
+require 'chef/resource'
+require 'chef/resource/file'
+require 'uuidtools'
+
+class Chef
+  class Provider
+    class Profile < Chef::Provider
+      include Chef::Mixin::Command
+      provides :profile, os: "darwin"
+      provides :config_profile, os: "darwin"
+
+      def whyrun_supported?
+        true
+      end
+
+      def load_current_resource
+        @current_resource = Chef::Resource::Profile.new(@new_resource.name)
+        @current_resource.profile_name(@new_resource.profile_name)
+        @cache_dir = Chef::FileCache.create_cache_path(
+          "profiles/#{@new_resource.cookbook_name}"
+        )
+        @cookbook_profile = nil
+
+        @all_profiles = get_installed_profiles
+        @new_profile = @new_resource.profile || @new_resource.profile_name
+
+        @new_profile_hash = get_new_profile_hash
+
+        if @new_profile_hash
+          @new_profile_identifier = @new_profile_hash['PayloadIdentifier'] ||
+            @new_resource.identifier ||
+            @new_resource.profile_name
+        else
+          @new_profile_identifier = @new_resource.identifier ||
+          @new_resource.profile_name
+        end
+
+        current_profile = @all_profiles['_computerlevel'].find {
+          |item| item['ProfileIdentifier'] ==
+             @new_profile_identifier
+        }
+        @current_resource.profile(current_profile)
+
+        @current_resource.installed(profile_installed?)
+      end
+
+      def define_resource_requirements
+        requirements.assert(:remove) do |a|
+          if @new_profile_identifier
+            a.assertion {
+              !@new_profile_identifier.nil? and
+              !@new_profile_identifier.end_with?('.mobileconfig') and
+              /^\w+(?:\.\w+)+$/.match(@new_profile_identifier)
+            }
+            a.failure_message Chef::Exceptions::UnsupportedAction, "when removing using the identifier attribute, it must match the profile identifier"
```

Comment:
`Chef::Exceptions::UnsupportedAction` is used when `resource.action` is not supported. These errors are more specific to profile, so I'd add a `Chef::Exceptions::Profile` exception or just use RuntimeError if we don't think we'd use profile in another provider.